design/search.md

```diff
@@ -113,4 +115,31 @@ Permissions:
+
+### Search & query-translation
+The search service in general operates by taking in queries and translating them to SQL, fetching results, and then formatting them for consumption. The return formats will be similar to our existing APIs, so I won't go into detail on those, but the query translation bears some additional detail.
+
+I'll use a shorthand to express the intent of the query, though it will be done in the actual service with a more verbose JSON-based format, so as to avoid spending too much time writing parsers for a custom format.
+
+First, a simple example. Imagine a query like `label:"*example.bam*" AND path_prefix:"/iplant/home/atreyu/"`, which is about what the current simple search does. That query is expressed as two clauses, one for the label and one for the path prefix, joined with AND. Elasticsearch, and our current search service, use a structure with three lists: ES calls them must, should, and must_not; our search service uses all, any, none; so, in that option, the two clauses would appear in the 'must' or 'all' list. Upon getting the query, the service would generate fragments of SQL from each, intended to be combined into a WHERE clause. In this case, that would be sufficient, though in other cases the service may need to be able to modify other pieces of a SELECT query such as joins, window functions, or CTEs. For the label clause, assuming our defaults expect case-insensitive search with wildcards on both sides, the fragment would be something like `label ILIKE '%example.bam%'`; for the path, case sensitivity would need to be turn on, so it would be `path LIKE '/iplant/home/atreyu/%'`. (Despite the leading wildcard, a trigram index can help with the label clause). The produced query would therefore look roughly like: `SELECT (...relevant columns to produce JSON output) FROM data (... any joins needed to produce the output) WHERE label ILIKE '%example.bam%' AND path LIKE '/iplant/home/atreyu/%';`.
```

Comment:
Atreyu 😆 One of our Sonora user stories (I think I'm remembering that term correctly)

Reply:
Yeah, he was our main persona for that, and I figured that would be a good name to reuse XD